File: backend/permissions.py
"""Permission checking utilities and decorators"""
from fastapi import HTTPException, Depends
from typing import List, Callable
from auth import get_current_user
from db import get_supabase
from functools import wraps

# Define all available permissions
class Permissions:
    # Dashboard
    VIEW_DASHBOARD = "view_dashboard"
    EDIT_DASHBOARD = "edit_dashboard"
    
    # Inventory
    VIEW_INVENTORY = "view_inventory"
    EDIT_INVENTORY = "edit_inventory"
    GENERATE_ORDERS = "generate_orders"
    
    # Employees
    VIEW_EMPLOYEES = "view_employees"
    EDIT_EMPLOYEES = "edit_employees"
    MANAGE_PERMISSIONS = "manage_permissions"
    
    # Schedule
    VIEW_SCHEDULE = "view_schedule"
    EDIT_SCHEDULE = "edit_schedule"
    GENERATE_SCHEDULE = "generate_schedule"
    SET_AVAILABILITY = "set_availability"
    
    # Financials
    VIEW_FINANCIALS = "view_financials"
    EDIT_FINANCIALS = "edit_financials"
    
    # Reminders
    VIEW_REMINDERS = "view_reminders"
    EDIT_REMINDERS = "edit_reminders"
    SET_REMINDERS = "set_reminders"
    
    # Business
    EDIT_BUSINESS = "edit_business"

# Default permissions for each role
ROLE_PERMISSIONS = {
    "admin": [
        # Admin has ALL permissions
        Permissions.VIEW_DASHBOARD,
        Permissions.EDIT_DASHBOARD,
        Permissions.VIEW_INVENTORY,
        Permissions.EDIT_INVENTORY,
        Permissions.GENERATE_ORDERS,
        Permissions.VIEW_EMPLOYEES,
        Permissions.EDIT_EMPLOYEES,
        Permissions.MANAGE_PERMISSIONS,
        Permissions.VIEW_SCHEDULE,
        Permissions.EDIT_SCHEDULE,
        Permissions.GENERATE_SCHEDULE,
        Permissions.SET_AVAILABILITY,
        Permissions.VIEW_FINANCIALS,
        Permissions.EDIT_FINANCIALS,
        Permissions.VIEW_REMINDERS,
        Permissions.EDIT_REMINDERS,
        Permissions.SET_REMINDERS,
        Permissions.EDIT_BUSINESS,
    ],
    "employee": [
        # Employee has limited default permissions
        Permissions.VIEW_DASHBOARD,
        Permissions.VIEW_INVENTORY,
        Permissions.VIEW_SCHEDULE,
        Permissions.SET_AVAILABILITY,
        Permissions.VIEW_REMINDERS,
        Permissions.SET_REMINDERS,
    ]
}

def get_user_permissions(user_id: str) -> List[str]:
    """Get all permissions for a user (role + custom)"""
    supabase = get_supabase()
    
    # Get user profile
    result = supabase.table("profiles")\
        .select("role, custom_permissions")\
        .eq("id", user_id)\
        .eq("is_active", True)\
        .single()\
        .execute()
    
    if not result.data:
        return []
    
    role = result.data.get("role", "employee")
    custom_permissions = result.data.get("custom_permissions", [])
    
    # Admin gets all permissions
    if role == "admin":
        return ROLE_PERMISSIONS["admin"]
    
    # Get role default permissions
    role_perms = set(ROLE_PERMISSIONS.get(role, []))
    
    # Add custom permissions
    custom_perms = set(custom_permissions) if isinstance(custom_permissions, list) else set()
    
    # Combine and return
    all_perms = role_perms.union(custom_perms)
    return list(all_perms)

def has_permission(user: dict, permission: str) -> bool:
    """Check if user has a specific permission"""
    # Admin always has permission
    if user.get("role") == "admin":
        return True
    
    # Get user's permissions
    user_permissions = get_user_permissions(user["user_id"])
    
    return permission in user_permissions

def require_permission(permission: str):
    """
    Dependency to require a specific permission for an endpoint.
    Usage: @router.get("/", dependencies=[Depends(require_permission(Permissions.VIEW_INVENTORY))])
    """
    async def check_permission(current_user: dict = Depends(get_current_user)):
        if not has_permission(current_user, permission):
            raise HTTPException(
                status_code=403,
                detail=f"Permission denied. Required permission: {permission}"
            )
        return current_user
    
    return check_permission

def require_any_permission(*permissions: str):
    """
    Dependency to require ANY of the listed permissions.
    Usage: dependencies=[Depends(require_any_permission(Permissions.EDIT_INVENTORY, Permissions.ADMIN))]
    """
    async def check_any_permission(current_user: dict = Depends(get_current_user)):
        user_perms = get_user_permissions(current_user["user_id"])
        
        if not any(perm in user_perms for perm in permissions):
            raise HTTPException(
                status_code=403,
                detail=f"Permission denied. Required one of: {', '.join(permissions)}"
            )
        return current_user
    
    return check_any_permission

def require_all_permissions(*permissions: str):
    """
    Dependency to require ALL of the listed permissions.
    Usage: dependencies=[Depends(require_all_permissions(Permissions.VIEW_INVENTORY, Permissions.EDIT_INVENTORY))]
    """
    async def check_all_permissions(current_user: dict = Depends(get_current_user)):
        user_perms = get_user_permissions(current_user["user_id"])
        
        if not all(perm in user_perms for perm in permissions):
            raise HTTPException(
                status_code=403,
                detail=f"Permission denied. Required all of: {', '.join(permissions)}"
            )
        return current_user
    
    return check_all_permissions

def require_admin():
    """Dependency to require admin role"""
    async def check_admin(current_user: dict = Depends(get_current_user)):
        if current_user.get("role") != "admin":
            raise HTTPException(
                status_code=403,
                detail="Admin access required"
            )
        return current_user
    
    return check_admin

File: frontend/src/lib/permissions.ts
/**
 * Permission constants and utilities
 * Matches backend permissions.py
 */

export const Permissions = {
  // Dashboard
  VIEW_DASHBOARD: 'view_dashboard',
  EDIT_DASHBOARD: 'edit_dashboard',
  
  // Inventory
  VIEW_INVENTORY: 'view_inventory',
  EDIT_INVENTORY: 'edit_inventory',
  GENERATE_ORDERS: 'generate_orders',
  
  // Employees
  VIEW_EMPLOYEES: 'view_employees',
  EDIT_EMPLOYEES: 'edit_employees',
  MANAGE_PERMISSIONS: 'manage_permissions',
  
  // Schedule
  VIEW_SCHEDULE: 'view_schedule',
  EDIT_SCHEDULE: 'edit_schedule',
  GENERATE_SCHEDULE: 'generate_schedule',
  SET_AVAILABILITY: 'set_availability',
  
  // Financials
  VIEW_FINANCIALS: 'view_financials',
  EDIT_FINANCIALS: 'edit_financials',
  
  // Reminders
  VIEW_REMINDERS: 'view_reminders',
  EDIT_REMINDERS: 'edit_reminders',
  SET_REMINDERS: 'set_reminders',
  
  // Business
  EDIT_BUSINESS: 'edit_business',
} as const

export type PermissionType = typeof Permissions[keyof typeof Permissions]

export interface PermissionInfo {
  name: string
  category: string
  description: string
  label: string
}

// Permission metadata for UI display
export const PERMISSION_INFO: Record<string, PermissionInfo> = {
  [Permissions.VIEW_DASHBOARD]: {
    name: Permissions.VIEW_DASHBOARD,
    category: 'Dashboard',
    description: 'View dashboard and statistics',
    label: 'View Dashboard'
  },
  [Permissions.EDIT_DASHBOARD]: {
    name: Permissions.EDIT_DASHBOARD,
    category: 'Dashboard',
    description: 'Edit dashboard settings',
    label: 'Edit Dashboard'
  },
  [Permissions.VIEW_INVENTORY]: {
    name: Permissions.VIEW_INVENTORY,
    category: 'Inventory',
    description: 'View inventory items',
    label: 'View Inventory'
  },
  [Permissions.EDIT_INVENTORY]: {
    name: Permissions.EDIT_INVENTORY,
    category: 'Inventory',
    description: 'Add, edit, and delete inventory items',
    label: 'Edit Inventory'
  },
  [Permissions.GENERATE_ORDERS]: {
    name: Permissions.GENERATE_ORDERS,
    category: 'Inventory',
    description: 'Generate AI-powered inventory orders',
    label: 'Generate Orders'
  },
  [Permissions.VIEW_EMPLOYEES]: {
    name: Permissions.VIEW_EMPLOYEES,
    category: 'Employees',
    description: 'View employee list',
    label: 'View Employees'
  },
  [Permissions.EDIT_EMPLOYEES]: {
    name: Permissions.EDIT_EMPLOYEES,
    category: 'Employees',
    description: 'Add, edit, and delete employees',
    label: 'Edit Employees'
  },
  [Permissions.MANAGE_PERMISSIONS]: {
    name: Permissions.MANAGE_PERMISSIONS,
    category: 'Employees',
    description: 'Manage user roles and permissions',
    label: 'Manage User Permissions'
  },
  [Permissions.VIEW_SCHEDULE]: {
    name: Permissions.VIEW_SCHEDULE,
    category: 'Schedule',
    description: 'View work schedules',
    label: 'View Schedule'
  },
  [Permissions.EDIT_SCHEDULE]: {
    name: Permissions.EDIT_SCHEDULE,
    category: 'Schedule',
    description: 'Create and edit schedules',
    label: 'Edit Schedule'
  },
  [Permissions.GENERATE_SCHEDULE]: {
    name: Permissions.GENERATE_SCHEDULE,
    category: 'Schedule',
    description: 'Generate AI-powered schedules',
    label: 'Generate AI Schedule'
  },
  [Permissions.SET_AVAILABILITY]: {
    name: Permissions.SET_AVAILABILITY,
    category: 'Schedule',
    description: 'Set own availability',
    label: 'Set Own Availability'
  },
  [Permissions.VIEW_FINANCIALS]: {
    name: Permissions.VIEW_FINANCIALS,
    category: 'Financials',
    description: 'View financial data',
    label: 'View Financials'
  },
  [Permissions.EDIT_FINANCIALS]: {
    name: Permissions.EDIT_FINANCIALS,
    category: 'Financials',
    description: 'Edit financial transactions',
    label: 'Edit Financials'
  },
  [Permissions.VIEW_REMINDERS]: {
    name: Permissions.VIEW_REMINDERS,
    category: 'Reminders',
    description: 'View reminders',
    label: 'View Reminders'
  },
  [Permissions.EDIT_REMINDERS]: {
    name: Permissions.EDIT_REMINDERS,
    category: 'Reminders',
    description: 'Create and edit reminders',
    label: 'Edit Reminders'
  },
  [Permissions.SET_REMINDERS]: {
    name: Permissions.SET_REMINDERS,
    category: 'Reminders',
    description: 'Set personal reminders',
    label: 'Set Personal Reminders'
  },
  [Permissions.EDIT_BUSINESS]: {
    name: Permissions.EDIT_BUSINESS,
    category: 'Business',
    description: 'Edit business settings and branding',
    label: 'Edit Business Settings'
  },
}

// Group permissions by category for UI display
export const getPermissionsByCategory = (): Record<string, PermissionInfo[]> => {
  const grouped: Record<string, PermissionInfo[]> = {}
  
  Object.values(PERMISSION_INFO).forEach(perm => {
    if (!grouped[perm.category]) {
      grouped[perm.category] = []
    }
    grouped[perm.category].push(perm)
  })
  
  return grouped
}
